common/task/interface.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"github.com/uber/cadence/common"
+)
+
+type (
+	// SequentialTaskProcessor is the generic coroutine pool interface
+	// which process sequential task
+	SequentialTaskProcessor interface {
+		common.Daemon
+		Submit(task SequentialTask) error
+	}
+
+	// Task is the generic task representation
+	Task interface {
+		// Execute process this task
+		Execute() error
+		// HandleErr handle the error returned by Execute
+		HandleErr(err error) error
+		// RetryErr check whether to retry after HandleErr(Execute())
+		RetryErr(err error) bool
+		// Ack marks the task as successful completed
+		Ack()
+		// Nack marks the task as unsuccessful completed
+		Nack()
+	}
+
+	// SequentialTask is the interface for tasks which should be executed sequentially
+	SequentialTask interface {
+		// QueueID return the ID of the queue this task belongs to
+		QueueID() interface{}
+		// TaskID return the ID of the task, this task ID is used for sorting
+		TaskID() int64
+		// HashCode return the hash code of the task
+		HashCode() uint32
```

Comment:
QueueID() and HashCode() need to satisfy certain properties and are closely related but this definition makes it look like they are simply task properties i.e. all tasks related to same workflow must produce same queueID and same hashCode is not obvious. And when queueIDs are equal, hashCode MUST be equal as well i.e must be Comparable, in java terminology

How about you specifically separating out the partition key part and adding comments about requirements for partition key
```
SequentialTaskPartitionKey interface {
    Key() interface // MUST be go primitive type or struct with primitive types 
    HashCode() uint32
}
```

And then having your SequentialTask interface like 
```
SequentialTask interface {
    TaskID() int64
     PartitionKey() SequentialTaskPartitionKey
}
```

Reply:
Alternatively, you can also consider something like this
```
type SequentialTaskPartitionKey definition.WorkflowIdentifier

func (p *SequentialTaskPartitionKey) HashCode() uint32 {...}

SequentialTask interface {
    TaskID() int64
    PartitionKey() SequentialTaskPartitionKey
}
 
```